### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data_df = pd.read_csv(file_to_load)
purchase_data_df

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price  
0    Extraction, Quickblade Of Trembling Hands   3.53  
1                            Frenzied Scimitar   1.56  
2                                 Final Critic   4.88  
3                                  Blindscythe   3.27  
4                                         Fury   1.44  
..                                         ...    ...  
775                                       Wolf   3.54  
776                           Exiled Doomblade   1.63  
777      Celeste, Incarnation of the Corrupted   3.46  
778                               Final Critic   4.19  
779                                       Dawn   4.60  

[780 rows x 7 columns]

In [2]:
# Determine if player names appear more than once
purchase_data_df.duplicated(subset=['SN'])

0      False
1      False
2      False
3      False
4      False
       ...  
775    False
776     True
777    False
778    False
779     True
Length: 780, dtype: bool

## Player Count

* Display the total number of players


In [3]:
# Create a DF that drops duplicate players (names)
no_duplicate_names_df = purchase_data_df.drop_duplicates(subset=['SN'])
no_duplicate_names_df["SN"].value_counts()

# Create a player count by taking the length of the new list of players with dropped duplicates
total_players = len(no_duplicate_names_df["SN"].value_counts())
print(f"Total players: {total_players}")

Total players: 576


In [4]:
# See if there are duplicate item names (applies to cell below)
purchase_data_df.duplicated(subset=['Item Name'])

0      False
1      False
2      False
3      False
4      False
       ...  
775     True
776     True
777     True
778     True
779     True
Length: 780, dtype: bool

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [5]:
# Create a DF that drops the duplicate items
no_duplicate_items_df = purchase_data_df.drop_duplicates(subset=['Item Name'])
no_duplicate_items_df["Item Name"].value_counts()

# Create an item count by taking the length of the new list of items with dropped duplicates
unique_items = len(no_duplicate_items_df["Item Name"].value_counts())
print(f"Unique items: {unique_items}")

Unique items: 179


In [6]:
# Number of purchases
purchase_data_df["Purchase ID"].value_counts()
num_purchases = len(purchase_data_df["Purchase ID"].value_counts())
print(f"Number of purchases: {num_purchases}")

Number of purchases: 780


In [7]:
# Total revenue
revenue = purchase_data_df["Price"].sum()
print(f"Total revenue: ${revenue}")

Total revenue: $2379.77


In [8]:
# Average price
avg_price = round(revenue/num_purchases, 2)
print(f"Average price: ${avg_price}")

Average price: $3.05


In [9]:
# Create a summary DF for purchases
purchasing_analysis_df = pd.DataFrame(
{"No. Unique Items": [unique_items],
 "No. Purchases": [num_purchases],
 "Total Revenue": [revenue],
 "Average Price": [avg_price]
})
purchasing_analysis_df

No. Unique Items  No. Purchases  Total Revenue  Average Price
0               179            780        2379.77           3.05

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [10]:
# Filter data so that only gender is in the DF
gender_df = no_duplicate_names_df["Gender"]

In [11]:
# Gender count
Gender_count = no_duplicate_names_df["Gender"].value_counts()

# Gender total
Gender_total = Gender_count.sum()

# Gender percentages
Gender_percent = (Gender_count/Gender_total)*100

In [12]:
# Create a summary DF for gender demographics
gender_df = pd.DataFrame ({"Total Count": Gender_count, 
                           "Percentage of Players": Gender_percent})

# Format the percentages so that it rounds to the second decimal place and adds a "%
gender_df["Percentage of Players"] = gender_df["Percentage of Players"].map("{:.2f}%".format)
gender_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [13]:
# Using GroupBy to separate the full purchase data into fields by gender
gender_purchases_df = purchase_data_df.groupby(['Gender'])

In [14]:
# Purchase count by gender
gender_purchase_count = gender_purchases_df["Purchase ID"].count()

In [15]:
# Total purchase value by gender
gender_total_purchases = gender_purchases_df["Price"].sum()

In [16]:
# Average purchase price by gender
gender_avgprice = round(gender_total_purchases/gender_purchase_count, 2)

In [17]:
# Using GroupBy to separate the data without duplicate players into fields by gender
noduplicates_gender = no_duplicate_names_df.groupby(['Gender'])

In [18]:
# Calculating avg. total purchase per person by taking the sum of purchases (price) from full DataFrame 
# and dividing it by the count of players from the DF without duplicate players

avg_tot_pers = gender_total_purchases / noduplicates_gender['SN'].count()

In [19]:
# Create a DF for purchasing analysis by gender

gender_df = pd.DataFrame ({"Total Count": gender_purchase_count,
                           "Average Purchase Price": gender_avgprice,
                           "Total Purchase Value": gender_total_purchases,
                          "Avg Total Purchase per Person": avg_tot_pers})
gender_df["Average Purchase Price"] = gender_df["Average Purchase Price"].map("${:.2f}".format)
gender_df["Total Purchase Value"] = gender_df["Total Purchase Value"].map("${:.2f}".format)
gender_df["Avg Total Purchase per Person"] = gender_df["Avg Total Purchase per Person"].map("${:.2f}".format)

gender_df

Total Count Average Purchase Price  \
Gender                                                      
Female                         113                  $3.20   
Male                           652                  $3.02   
Other / Non-Disclosed           15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [20]:
# Find the age range
age_min = no_duplicate_names_df['Age'].min()
age_max = no_duplicate_names_df['Age'].max()
print(f"Minimum age: {age_min}")
print(f"Maximum age: {age_max}")

Minimum age: 7
Maximum age: 45


In [21]:
# Create bins
bins = [0, 9.90 , 14.90, 19.90 , 24.90, 29.90, 34.90, 39.90, 99999]
age_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"] 

In [22]:
# Slice the data and place it into bins [using DF without duplicate players]
pd.cut(no_duplicate_names_df["Age"], bins, labels=age_labels)

0      20-24
1        40+
2      20-24
3      20-24
4      20-24
       ...  
773    20-24
774    10-14
775    20-24
777    20-24
778      <10
Name: Age, Length: 576, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [23]:
# Place the data series into a new column inside of the DF
no_duplicate_names_df["Age Range"] = pd.cut(no_duplicate_names_df["Age"], bins, labels=age_labels)

/opt/anaconda3/envs/PythonData/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [24]:
# Create DF with only age range
age_df = no_duplicate_names_df["Age Range"]

In [25]:
# Age count
Age_count = no_duplicate_names_df["Age Range"].value_counts()

# Age total
Age_total = Age_count.sum()

# Age percentages
Age_percent = (Age_count/Age_total)*100

In [26]:
# Create a summary DF for age demographics
age_df = pd.DataFrame ({"Total Count": Age_count, "Percentage of Players": Age_percent})

# Format the percentages so that it rounds to the second decimal place and adds a "%
age_df["Percentage of Players"] = age_df["Percentage of Players"].map("{:.2f}%".format)
age_df = age_df.sort_index()
age_df

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [27]:
# Create bins for the full purchase DF
bins = [0, 9.90 , 14.90, 19.90 , 24.90, 29.90, 34.90, 39.90, 99999]
age_lables = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"] 

In [28]:
# Slice the data and place it into bins [Using full purchase DF]
pd.cut(purchase_data_df["Age"], bins, labels=age_labels)

0      20-24
1        40+
2      20-24
3      20-24
4      20-24
       ...  
775    20-24
776    20-24
777    20-24
778      <10
779    20-24
Name: Age, Length: 780, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [29]:
# Place the data series into a new column inside of the DF
purchase_data_df["Age Range"] = pd.cut(purchase_data_df["Age"], bins, labels=age_labels)

In [30]:
# Using GroupBy to separate the full purchase data into fields by age range
age_purchases_df = purchase_data_df.groupby(['Age Range'])

In [31]:
# Purchase count by age range
age_purchase_count = age_purchases_df["Purchase ID"].count()

In [32]:
# Total purchase value by age range
age_total_purchases = age_purchases_df["Price"].sum()

In [33]:
# Average purchase price by age range
age_avgprice = round(age_total_purchases/age_purchase_count, 2)

In [34]:
# Using GroupBy to separate the data without duplicate players into fields by age range
noduplicates_age = no_duplicate_names_df.groupby(['Age Range'])

In [35]:
# Calculating avg. total purchase per person by taking the sum of purchases (price) from the full DF
# and dividing it by the count of players from the DF without duplicate players

avg_tot_age = age_total_purchases / noduplicates_age['SN'].count()

In [36]:
# Create a DF for purchasing analysis by age range
age_df = pd.DataFrame ({"Total Count": age_purchase_count,
                           "Average Purchase Price": age_avgprice,
                           "Total Purchase Value": age_total_purchases,
                          "Avg Total Purchase per Person": avg_tot_age})
age_df["Average Purchase Price"] = age_df["Average Purchase Price"].map("${:.2f}".format)
age_df["Total Purchase Value"] = age_df["Total Purchase Value"].map("${:.2f}".format)   
age_df["Avg Total Purchase per Person"] = age_df["Avg Total Purchase per Person"].map("${:.2f}".format)
age_df

Total Count Average Purchase Price Total Purchase Value  \
Age Range                                                            
<10                 23                  $3.35               $77.13   
10-14               28                  $2.96               $82.78   
15-19              136                  $3.04              $412.89   
20-24              365                  $3.05             $1114.06   
25-29              101                  $2.90              $293.00   
30-34               73                  $2.93              $214.00   
35-39               41                  $3.60              $147.67   
40+                 13                  $2.94               $38.24   

          Avg Total Purchase per Person  
Age Range                                
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [37]:
# Group by player name
spenders_df = purchase_data_df.groupby(['SN'])

In [38]:
# Purchase count by name
spender_purchase_count = spenders_df["Purchase ID"].count()

In [39]:
# Total purchase value by name
spender_purchase_total = spenders_df["Price"].sum()

In [40]:
# Average purchase price by name
spender_avg_purchase_price = spenders_df["Price"].mean()

In [41]:
# Create a summary DF for spenders & sort so that values are descending
top_spenders_df = pd.DataFrame({"Purchase Count": spender_purchase_count,
                             "Average Purchase Price": spender_avg_purchase_price,
                             "Total Purchase Value": spender_purchase_total})
top_spenders_df = top_spenders_df.sort_values("Total Purchase Value",ascending=False)

# Format to round and add "$""
top_spenders_df["Average Purchase Price"] = top_spenders_df["Average Purchase Price"].map("${:.2f}".format)
top_spenders_df["Total Purchase Value"] = top_spenders_df["Total Purchase Value"].map("${:.2f}".format)      
top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [42]:
# Group by item name
popular_df = purchase_data_df.groupby(['Item ID', "Item Name"])
popular_df.count().head()

Purchase ID  SN  Age  Gender  Price  Age Range
Item ID Item Name                                                         
0       Splinter                      4   4    4       4      4          4
1       Crucifer                      4   4    4       4      4          4
2       Verdict                       6   6    6       6      6          6
3       Phantomlight                  6   6    6       6      6          6
4       Bloodlord's Fetish            5   5    5       5      5          5

In [43]:
# Purchase count by item id and name
popular_purchase_count = popular_df['Purchase ID'].count()

In [44]:
# Total purchase value by item id and name
popular_price = popular_df['Price'].sum()

In [45]:
# Average purchase price by item id and name
item_price = popular_df['Price'].mean()

In [46]:
# Create a summary DF for popular items & sort so that values are descending
popular_items_df = pd.DataFrame({"Purchase Count": popular_purchase_count,
                             "Item Price": item_price,
                             "Total Purchase Value": popular_price})

popular_items_df = popular_items_df.sort_values("Purchase Count",ascending=False)

# Format to round and add "$"
popular_items_df["Item Price"] = popular_items_df["Item Price"].map("${:.2f}".format)
popular_items_df["Total Purchase Value"] = popular_items_df["Total Purchase Value"].map("${:.2f}".format)      
popular_items_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

# HEROES OF PYMOLI PURCHASING ANALYSIS, FINAL REPORT

Overall, there were 179 unique items purchased by players, with a total of 780 purchases made. The total revenue was 
just over two thouand and three hundred dollars, and the average amount of money spent by players was around three dollars.

A total of 576 unique players made purchases. Of these, over eighty percent of players were male, were female, and identified as “other” or did not disclose their gender. Overall, male players generated the most revenue as compared to females and those who identified as another gender or did not disclose their gender. This gender difference in purchase revenue may be due to the overall larger number of male players. When looking at the amount of money spent by players at the individual level, gender is less of an influence. That is, regardless of gender, players seem to spend similar amounts of money when looking a at average purchase price and average total purchase per person.

Players’ ages ranged from 7 to 45 years. The largest age group represented was 20-24-year olds. Over 40 percent of players fell into this age range. The next two largest age range groups were 15-19-year old’s and 25-29-year olds, respectively. A small portion of players fell into the 0 to 9 age range. Overall, players who were 20 to 24 years generated the most revenue and players who were 40+ years old generated the least amount of revenue. When looking more closely at individuals’ purchasing behaviors, age has a smaller influence. Players had fairly similar average purchase prices. Although, there was more variability in the average total purchase price per person based on age. Interestingly, the highest average total purchase price was for players who were 35 to 39 years old, followed by those who were 0 to 9 years old. Players who were 40+ years old had the lowest average total purchase prices. 

The top five most popular items according to purchase count are “Final Critic”, “Oathbreaker, Last Hope of the Breaking Storm”, “Fiery Glass Crusader”, “Persuasion”, and “Extraction, Quickblade of Trembling Hands”. The top five more profitable items according to the total revenue they generated were: “Final Critic”, “Oathbreaker, Last Hope of the Breaking Storm”, “Nirvana”, “Fiery Glass Crusader”, and “Singed Scalpel”. Thus, to some extent there is some overlap in that the most popular items generate the most money. This is true for “Final Critic”, “Oathbreaker, Last Hope of the Breaking Storm”, and “Fiery Glass Crusader”. However, there are also popular items that while being purchased often, do not generate the top revenue, i.e., “Persuasion” and “Extraction, Quickblade of Trembling Hands”. Similarly, there are some items that are highly profitable, but are not the most popular according to purchase counts, i.e., “Nirvana” and “Singed” Scalpel”.

Overall, the insights drawn are: (1) This company should attempt to attract more players from the 0 to 9 and 35 to 39 age ranges. By doing so they would likely generate more total revenue as these players seem to be spending more money on average but are currently under-represented compared to those in their twenties; (2) Similarly, the company should attempt to attract more female players and other individuals who do not identify as “males”.  While players spend similar amounts of money on individual purchases, the company would significantly increase their total revenue by attracting larger numbers of players who do not identify as male, as non-males are underrepresented.